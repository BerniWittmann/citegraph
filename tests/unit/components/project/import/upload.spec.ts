import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../setupPlugins'

import ProjectImportUploadComponent from '@/components/project/import/Upload.vue'

jest.useFakeTimers()

describe('components/project/import/Upload.vue', () => {
  function getWrapper () {
    return shallowMount(ProjectImportUploadComponent, {
      i18n,
      stubs: {
        'vue-dropzone': '<div class="dropzone"><slot/></div>'
      }
    })
  }

  it('renders', () => {
    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the dropzone', () => {
    const wrapper = getWrapper()
    expect(wrapper.contains('#dropzone')).toBeTruthy()
    expect(wrapper.contains('#dropzone-previews')).toBeTruthy()
  })

  it('renders the import button only if there exist files', () => {
    const wrapper = getWrapper()
    const dz = wrapper.find('#dropzone')
    expect(wrapper.contains('v-btn-stub')).toBeFalsy()
    dz.vm.$emit('vdropzone-file-added')

    const button = wrapper.find('v-btn-stub')
    expect(button.exists()).toBeTruthy()
    expect(button.text()).toContain('project.import.upload.import')
  })

  it('processes the upload queue when the import button is clicked', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getQueuedFiles = jest.fn().mockReturnValue([{}])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getAcceptedFiles = jest.fn().mockReturnValue([{}])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.processQueue = jest.fn()
    const dz = wrapper.find('#dropzone')
    dz.vm.$emit('vdropzone-file-added')

    const button = wrapper.find('v-btn-stub')
    expect(button.exists()).toBeTruthy()
    button.vm.$emit('click')

    // @ts-ignore
    expect(wrapper.vm.$refs.myVueDropzone.processQueue).toHaveBeenCalled()
  })

  it('directly goes to the next step when there are no queued but accepted files', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getQueuedFiles = jest.fn().mockReturnValue([])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getAcceptedFiles = jest.fn().mockReturnValue([{}])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.processQueue = jest.fn()
    const dz = wrapper.find('#dropzone')
    dz.vm.$emit('vdropzone-file-added')
    expect(wrapper.emitted('next-step')).toBeFalsy()

    const button = wrapper.find('v-btn-stub')
    expect(button.exists()).toBeTruthy()
    button.vm.$emit('click')

    // @ts-ignore
    expect(wrapper.vm.$refs.myVueDropzone.processQueue).not.toHaveBeenCalled()
    expect(wrapper.emitted('next-step')).toBeTruthy()
  })

  it('does not trigger the upload if there are no files at all', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getQueuedFiles = jest.fn().mockReturnValue([])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getAcceptedFiles = jest.fn().mockReturnValue([])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.processQueue = jest.fn()
    const dz = wrapper.find('#dropzone')
    dz.vm.$emit('vdropzone-file-added')
    expect(wrapper.emitted('next-step')).toBeFalsy()

    const button = wrapper.find('v-btn-stub')
    expect(button.exists()).toBeTruthy()
    button.vm.$emit('click')

    // @ts-ignore
    expect(wrapper.vm.$refs.myVueDropzone.processQueue).not.toHaveBeenCalled()
    expect(wrapper.emitted('next-step')).toBeFalsy()
  })

  it('emits the next step event when the upload finished', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getQueuedFiles = jest.fn().mockReturnValue([{}])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.getAcceptedFiles = jest.fn().mockReturnValue([{}])
    // @ts-ignore
    wrapper.vm.$refs.myVueDropzone.processQueue = jest.fn()
    const dz = wrapper.find('#dropzone')
    dz.vm.$emit('vdropzone-file-added')
    const button = wrapper.find('v-btn-stub')
    expect(button.exists()).toBeTruthy()
    button.vm.$emit('click')
    expect(wrapper.emitted('next-step')).toBeFalsy()

    expect(button.html()).toMatchSnapshot()

    dz.vm.$emit('vdropzone-success-multiple')
    expect(button.html()).toMatchSnapshot()
    jest.runAllTimers()
    expect(wrapper.emitted('next-step')).toBeTruthy()
  })

  it('provides the file preview template', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    expect(wrapper.vm.isLoading).toBeFalsy()

    // @ts-ignore
    expect(wrapper.vm.previewHTML).toEqual(expect.any(String))
  })
})
