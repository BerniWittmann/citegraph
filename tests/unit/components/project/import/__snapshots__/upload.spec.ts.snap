// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/project/import/Upload.vue emits the next step event when the upload finished 1`] = `
<v-btn-stub color="primary" tag="button" activeclass="" xlarge="true" type="button" class="mr-10" loading="true">
  <v-icon-stub>mdi-play</v-icon-stub>
  project.import.upload.import
</v-btn-stub>
`;

exports[`components/project/import/Upload.vue emits the next step event when the upload finished 2`] = `
<v-btn-stub color="success" tag="button" activeclass="" xlarge="true" type="button" class="mr-10">
  <v-icon-stub>mdi-check</v-icon-stub>
  project.import.upload.import
</v-btn-stub>
`;

exports[`components/project/import/Upload.vue renders 1`] = `
<div class="project-import-upload">
  <h1>project.import.upload.title</h1>
  <v-container-stub tag="div" fluid="true" class="mb-4">
    <v-row-stub tag="div" class="mb-4">
      <v-col-stub tag="div" class="mr-4">
        <div class="dropzone" id="dropzone" options="[object Object]" use-custom-slot="" duplicate-check="">
          <v-hover-stub opendelay="0" closedelay="0" class="mx-3"></v-hover-stub>
        </div>
      </v-col-stub>
      <v-col-stub tag="div" class="ml-4">
        <div id="dropzone-previews"></div>
      </v-col-stub>
    </v-row-stub>
    <v-row-stub tag="div">
      <v-spacer-stub></v-spacer-stub>
      <!---->
    </v-row-stub>
  </v-container-stub>
</div>
`;
